// cached-constructors-x/dist/cached-constructors-x.esm.js
import noop from 'noop-x';
/**
 * Cached Array constructor.
 *
 * @class Array
 */

export var ArrayCtr = [].constructor;
/**
 * Cached Array prototype.
 *
 * @type {!object}
 */

export var ArrayPrototype = ArrayCtr.prototype;
/**
 * Cached Array constructor.
 *
 * @class Boolean
 */

export var BooleanCtr = true.constructor;
/**
 * Cached Boolean prototype.
 *
 * @type {!object}
 */

export var BooleanPrototype = BooleanCtr.prototype;
/**
 * Cached Function constructor.
 *
 * @class Function
 */

export var FunctionCtr = noop.constructor;
/**
 * Cached Function prototype.
 *
 * @type {!object}
 */

export var FunctionPrototype = FunctionCtr.prototype;
/**
 * Cached Number constructor.
 *
 * @class Number
 */

export var NumberCtr = 0 .constructor;
/**
 * Cached Number prototype.
 *
 * @type {!object}
 */

export var NumberPrototype = NumberCtr.prototype;
/**
 * Cached Object constructor.
 *
 * @class Object
 */

export var ObjectCtr = {}.constructor;
/**
 * Cached Object prototype.
 *
 * @type {!object}
 */

export var ObjectPrototype = ObjectCtr.prototype;
/**
 * Cached RegExp constructor.
 *
 * @class RegExp
 */

export var RegExpCtr = /none/.constructor;
/**
 * Cached RegExp prototype.
 *
 * @type {!object}
 */

export var RegExpPrototype = RegExpCtr.prototype;
/**
 * Cached String constructor.
 *
 * @class String
 */

export var StringCtr = ''.constructor;
/**
 * Cached String prototype.
 *
 * @type {!object}
 */

export var StringPrototype = StringCtr.prototype;
